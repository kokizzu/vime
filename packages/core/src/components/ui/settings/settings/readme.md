# vime-settings

A container for a collection of submenus and options for the player. On desktop, the settings is 
displayed as a small popup menu (scroll appears if `height >= maxHeight`) on the bottom right-hand
side of a video player, or slightly above the right-hand side of an audio player. On mobile, 
the settings is displayed as a [bottom sheet](https://material.io/components/sheets-bottom).

## Example

```htm
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-settings>
      <vime-submenu label="Playback Rate" hint="Normal">
        <vime-menu-radio-group value="1">
          <vime-menu-radio label="0.5" value ="0.5" />
          <vime-menu-radio label="Normal" value ="1" />
          <vime-menu-radio label="2" value ="2" />
        </vime-menu-radio-group>
      </vime-submenu>

      <vime-submenu label="Playback Quality" hint="720p">
        <vime-menu-radio-group value="720p">
          <vime-menu-radio label="1080p" value ="1080p" />
          <vime-menu-radio label="720p" value ="720p" />
          <vime-menu-radio label="480p" value ="480p" />
        </vime-menu-radio-group>
      </vime-submenu>
    </vime-settings>
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Properties

| Property         | Attribute         | Description                                                                                            | Type      | Default |
| ---------------- | ----------------- | ------------------------------------------------------------------------------------------------------ | --------- | ------- |
| `active`         | `active`          | Whether the settings menu is opened/closed.                                                            | `boolean` | `false` |
| `controlsHeight` | `controls-height` | The height of any lower control bar in pixels so that the settings can re-position itself accordingly. | `number`  | `65`    |


## Methods

### `setController(id: string, controller: SettingsController) => Promise<void>`

Sets the controller responsible for opening/closing this settings.

#### Returns

Type: `Promise<void>`




## Slots

| Slot | Description                                                                     |
| ---- | ------------------------------------------------------------------------------- |
|      | Used to pass in the body of the settings menu, which usually contains submenus. |


## CSS Custom Properties

| Name                            | Description                                                  |
| ------------------------------- | ------------------------------------------------------------ |
| `--settings-audio-max-height`   | The max height of the settings when outside an audio player. |
| `--settings-border-radius`      | The border radius of the settings menu.                      |
| `--settings-max-height`         | The max height of the settings menu.                         |
| `--settings-padding`            | The padding inside the settings menu.                        |
| `--settings-scroll-thumb-color` | The color of the settings scrollbar thumb.                   |
| `--settings-scroll-track-color` | The color of the settings scrollbar track.                   |
| `--settings-scroll-width`       | The width of the settings scrollbar.                         |
| `--settings-shadow`             | The shadow cast around the settings menu frame.              |
| `--settings-width`              | The width of the settings menu.                              |


## Dependencies

### Used by

 - [vime-default-settings](../default-settings)

### Depends on

- [vime-menu](../menu)

### Graph
```mermaid
graph TD;
  vime-settings --> vime-menu
  vime-default-settings --> vime-settings
  style vime-settings fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
