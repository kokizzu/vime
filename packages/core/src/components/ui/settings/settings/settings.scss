vime-settings {
  /**
   * @prop --settings-width: The width of the settings menu.
   * @prop --settings-padding: The padding inside the settings menu.
   * @prop --settings-shadow: The shadow cast around the settings menu frame.
   * @prop --settings-border-radius: The border radius of the settings menu.
   * @prop --settings-max-height: The max height of the settings menu.
   * @prop --settings-scroll-width: The width of the settings scrollbar.
   * @prop --settings-scroll-thumb-color: The color of the settings scrollbar thumb.
   * @prop --settings-scroll-track-color: The color of the settings scrollbar track.
   */

  display: none;
  position: absolute;
  right: 8px;
  pointer-events: auto;
  overflow-x: hidden;
  overflow-y: auto;
  max-height: var(--settings-max-height);
  border-radius: var(--settings-border-radius);
  width: var(--settings-width);
  padding: var(--settings-padding);
  box-shadow: var(--settings-shadow);
  z-index: $z-index-menu;
  scrollbar-width: thin;
  scroll-behavior: smooth;
  scrollbar-color: var(--settings-scroll-thumb-color) var(--settings-scroll-track-color);

  &::-webkit-scrollbar {
    width: var(--settings-scroll-width);
  }

  &::-webkit-scrollbar-track {
    background: var(--settings-scroll-track-color);
  }

  &::-webkit-scrollbar-thumb {
    border-radius: var(--settings-scroll-width);
    background-color: var(--settings-scroll-thumb-color);
    border: 2px solid var(--menu-bg);
  }

  &.active {
    display: block;
  }
}