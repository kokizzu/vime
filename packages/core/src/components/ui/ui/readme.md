# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop
is `true`, and the player is not in fullscreen mode.

<!-- Auto Generated Below -->

## Usage

### Angular

```html {3-5} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### Html

```html {3-5}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
  </vime-ui>
</vime-player>
```

### React

```tsx {2,8}
import React from 'react';
import { VimePlayer, VimeUi } from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>{/* ... */}</VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {4-6,11,16} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import { VimePlayer, VimeUi } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
    },
  };
</script>
```

## Dependencies

### Used by

- [vime-default-ui](../default-ui)

### Graph

```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
