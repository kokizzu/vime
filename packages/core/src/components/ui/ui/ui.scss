vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}