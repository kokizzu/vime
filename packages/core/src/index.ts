import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, withCustomPlayerContext } from './components/core/player/PlayerContext';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  withPlayerContext,
  withCustomPlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
