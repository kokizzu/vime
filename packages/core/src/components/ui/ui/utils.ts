import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
