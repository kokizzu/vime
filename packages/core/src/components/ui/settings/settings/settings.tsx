/* eslint-disable no-param-reassign */

import {
  h, Host, Component, Prop, Watch, Method, Element, State,
} from '@stencil/core';
import { Disposal } from '../../../core/player/Disposal';
import { listen } from '../../../../utils/dom';
import { isUndefined, isNull } from '../../../../utils/unit';
import { SettingsController } from './SettingsController';
import { PlayerStateDispatcher, createPlayerStateDispatcher } from '../../../core/player/PlayerState';
import { PlayerProp, PlayerProps } from '../../../core/player/PlayerProp';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';

let idCount = 0;

/**
 * @slot - Used to pass in the body of the settings menu, which usually contains submenus.
 */
@Component({
  tag: 'vime-settings',
  styleUrl: 'settings.scss',
})
export class Settings {
  private id!: string;

  private menu!: HTMLVimeMenuElement;

  private disposal = new Disposal();

  private controller?: SettingsController;

  private dispatch!: PlayerStateDispatcher;

  @Element() el!: HTMLVimeSettingsElement;

  @State() controllerId?: string;

  /**
   * The height of any lower control bar in pixels so that the settings can re-position itself
   * accordingly.
   */
  @Prop() controlsHeight = 65;

  /**
   * Whether the settings menu is opened/closed.
   */
  // eslint-disable-next-line @stencil/strict-mutable
  @Prop({ mutable: true, reflect: true }) active = false;

  @Watch('active')
  onActiveChange() {
    this.dispatch(PlayerProp.IsSettingsActive, this.active);
    if (isUndefined(this.controller)) return;
    this.controller!.expanded = this.active;
  }

  /**
   * @internal
   */
  @Prop() isMobile: PlayerProps[PlayerProp.IsMobile] = false;

  /**
   * @internal
   */
  @Prop() isAudioView: PlayerProps[PlayerProp.IsAudioView] = false;

  connectedCallback() {
    this.dispatch = createPlayerStateDispatcher(this);
  }

  componentWillLoad() {
    idCount += 1;
    this.id = `vime-settings-${idCount}`;
  }

  disconnectedCallback() {
    this.disposal.empty();
  }

  /**
   * Sets the controller responsible for opening/closing this settings.
   */
  @Method()
  async setController(id: string, controller: SettingsController) {
    this.controllerId = id;
    this.controller = controller;
    this.controller.menu = this.id;
    this.disposal.add(listen(this.controller, 'click', () => { this.active = !this.active; }));
    this.disposal.add(listen(this.controller, 'keydown', (event: KeyboardEvent) => {
      // Enter (13)
      if (event.keyCode !== 13) return;
      // We're looking for !active because the `click` event above will toggle it to active.
      if (!this.active) this.menu.focusOnOpen();
    }));
  }

  private onClose(event: CustomEvent<void>) {
    if (isNull(event.target) || (event.target! as HTMLElement).id !== this.id) return;
    this.active = false;
  }

  render() {
    return (
      <Host
        style={{
          bottom: `${this.controlsHeight + 8}px`,
        }}
        class={{
          active: this.active,
          mobile: this.isMobile,
          audio: this.isAudioView,
        }}
      >
        <vime-menu
          identifier={this.id}
          active={this.active}
          controller={this.controllerId ?? ''}
          onClose={this.onClose.bind(this)}
          ref={(el: any) => { this.menu = el; }}
        >
          <slot />
        </vime-menu>
      </Host>
    );
  }
}

openPlayerWormhole(Settings, [
  PlayerProp.IsMobile,
  PlayerProp.IsAudioView,
]);
