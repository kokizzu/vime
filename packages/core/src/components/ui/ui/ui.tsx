import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.Playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.IsFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.IsVideoView,
  PlayerProp.Playsinline,
  PlayerProp.IsFullscreenActive,
]);
