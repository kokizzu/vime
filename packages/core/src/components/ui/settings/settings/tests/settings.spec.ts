import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../../ui/tests';
import { Settings } from '../settings';
import { SettingsController } from '../SettingsController';

let page: SpecPage;
let settings: HTMLVimeSettingsElement;
let menu: HTMLVimeMenuElement;
let controller: SettingsController;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [Settings],
    `
    <vime-settings>
     DefaultSlot 
    </vime-settings>
    `,
  ));

  settings = page.root!.querySelector('vime-settings')!;
  menu = page.root!.querySelector('vime-settings > vime-menu')! as HTMLVimeMenuElement;

  controller = {
    menu: undefined,
    expanded: false,
    addEventListener: jest.fn(),
  } as any;

  await settings.setController('controller', controller);
});

it('should be structurally sound', () => {
  expect(page.root).toMatchSnapshot();
});

it('should set the controller', async () => {
  await page.waitForChanges();
  expect(menu.getAttribute('controller')).toEqual('controller');
  expect(controller.menu).toEqual('vime-settings-2');
});

it('should readjust position based on controls height', async () => {
  settings.controlsHeight = 100;
  await page.waitForChanges();
  expect(settings.style.bottom).toEqual('108px');
});

it('should change active state', async () => {
  settings.active = true;
  await page.waitForChanges();
  expect(page.root!.isSettingsActive).toBeTruthy();
  expect(controller.expanded).toBeTruthy();
  settings.active = false;
  await page.waitForChanges();
  expect(page.root!.isSettingsActive).toBeFalsy();
  expect(controller.expanded).toBeFalsy();
});
