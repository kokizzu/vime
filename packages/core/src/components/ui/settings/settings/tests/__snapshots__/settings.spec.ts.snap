// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-player aria-hidden="true" aria-label="Video Player" id="vime-player-1" tabindex="0">
  <!---->
  <div class="active skeleton"></div>
  <vime-faketube></vime-faketube>
  <vime-ui>
    <!---->
    <div>
      <vime-settings style="bottom: 73px;">
        <!---->
        <vime-menu controller="" identifier="vime-settings-1">
          DefaultSlot
        </vime-menu>
      </vime-settings>
    </div>
  </vime-ui>
</vime-player>
`;
